### PR TITLE
Typescript Support - add declaration files and add fixes

### PR DESCRIPTION
**Pull request by mocca102**

### CODE REVIEW

File: src/types/types.d.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+}
+
+export interface ItemData extends Record<string, any> {
+	keywords?: string[];
+	url?: string;
+	image_url?: string;
+	facets?: Record<string, any>;
+	group_ids?: string[];
+	description?: string;
+	active?: boolean;
```
Comment (esezen) [inline src/types/types.d.ts:153]: Should we add `[key: string]: any;` here or is that covered somewhere else?

**Review by esezen — COMMENTED** [review]
Not sure if it's just me but it looks like some files use tabs and some files use spaces for indentation. Can we have all of them use spaces?

File: src/types/types.d.ts (hunk 1 above)
Comment (deomsj) [reply src/types/types.d.ts:153]: yup, this same functionality is accomplished with the expression:
`extends Record<string, any>` on line 146

More on the `Record` Utility Type:
https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type

**Review by deomsj — CHANGES_REQUESTED** [review]
Yes! Nice work on all this @mocca102!!! 

Again, this `expectType`  from the `tsd` package is so 💪🏻💪🏻💪🏻.

Doubling down on @esezen's comment around consistent src formatting. Personally, I like [prettier](https://prettier.io/), but there are other options available to ensure consistent formatting. Holla if you would like any support with this. 

Feel free to hit me up for another review when we've got our code formatting flushed out

**Review by deomsj — CHANGES_REQUESTED** [review]
Damn, this is 🔥🔥🔥 Thanks for the quick formatting fix and all the great work done here, @mocca102.

I left some inline comments with screenshots for a few quick fixes.

Requesting two high level changes:
1. Seeing all these types with their corresponding tests has me craving:
- 'catalog.test-d.ts'
- 'constructorio.test-d.ts'
- 'quizzes.test-d.ts'
- 'tasks.test-d.ts'
- 'tracker.test-d.ts'

^ I'd feel much more confident shipping their corresponding types if we had these test-d.ts files

2. Looks like we have a merge conflict with master's `package.json`, let's back merge master into this feature branch to resolve this merge conflict here.

**Comment by mocca102** [comment]
> 
> 1. Seeing all these types with their corresponding tests has me craving:
> 
> * 'catalog.test-d.ts'
> * 'constructorio.test-d.ts'
> * 'quizzes.test-d.ts'
> * 'tasks.test-d.ts'
> * 'tracker.test-d.ts'
> 
> ^ I'd feel much more confident shipping their corresponding types if we had these test-d.ts files
> 

Thanks a 100 for your feedback ❤️ 

Comments addressed. Didn't add tests only for tracker and constructorio since these don't have response examples types that I use to test against @deomsj

**Review by deomsj — APPROVED** [review]
Nice work @mocca102! Code looks great and this looks ready to launch to me. We may have some quick, small follow-up PRs needed to iterate on these types a bit, but this is passing all my tests and with the supportive, non-blocking patterns in place like `extends Record<string, any>` I believe this will already start adding a lot of value in current state 

---

Before merging anything, please double check me on these most recent changes, @mocca102 🙏🏻 

Note: I pushed two small commits to address outstanding issues surfaced by `npm run tes:types` taking us from this:
Screenshot 2022-12-15 at 2 25 27 PM

to this:
Screenshot 2022-12-15 at 2 27 50 PM

**Comment by mocca102** [comment]
> Nice work @mocca102! Code looks great and this looks ready to launch to me. We may have some quick, small follow-up PRs needed to iterate on these types a bit, but this is passing all my tests and with the supportive, non-blocking patterns in place like `extends Record<string, any>` I believe this will already start adding a lot of value in current state
> 
> Before merging anything, please double check me on these most recent changes, @mocca102 🙏🏻
> 
> Note: I pushed two small commits to address outstanding issues surfaced by `npm run tes:types` taking us from this: Screenshot 2022-12-15 at 2 25 27 PM
> 
> to this: Screenshot 2022-12-15 at 2 27 50 PM

This is ❤️ ❤️ ❤️

File: src/types/tracker.d.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+      item_is_convertible?: string;
+      section?: string;
+    },
+    userParameters?: TrackerUserParameters,
+    networkParameters?: NetworkParameters
+  ): true | Error;
+
+  trackConversion(
+    term?: string,
+    parameters: {
```
[the remark this answers is not in the document]
Comment (deomsj) [reply src/types/tracker.d.ts:96]: so is `term` a required field then? if someone wanted to pass parameters and no term, how could they do that?

[the remark this answers is not in the document]
Comment (mocca102) [reply src/types/tracker.d.ts:96]: something like that `trackConversion('', parameters)` or `trackConversion(undefined, parameters)`

File: src/types/tracker.d.ts [hunk 3]
```diff
@@ -101,6 +100,7 @@ declare class Tracker {
       result_id?: string;
       section?: string;
     },
+    term?: string,
```
[the remark this answers is not in the document]
Comment (deomsj) [reply src/types/tracker.d.ts]: yup, makes perfect sense. good catch :)

**Review by esezen — APPROVED** [review]
LGTM!

**Pull request merged**